### DISCUSSION TITLE
What's the rationale for the highlight blend behaviour?

### DISCUSSION THREAD

**Discussion by fredizzimo** [neovim/neovim#31737] [Q&A]
The documentation here https://neovim.io/doc/user/syntax.html#highlight-blend, says this
> Override the blend level for a highlight group within the popupmenu
	or floating windows. Only takes effect if ['pumblend'](https://neovim.io/doc/user/options.html#'pumblend') or ['winblend'](https://neovim.io/doc/user/options.html#'winblend')
	is set for the menu or window. See the help at the respective option.

I have a few questions regarding this.

1. Why does it only take effect when `pumblend`  or `winblend` is set? I see very little reason for that behaviour, and it makes it much harder to implement a floating window with only a very small see through part for example. I think a more sensible specification would be that the highlight blend unconditionally overwrites `winblend` and `pumblend` so that they specify the defaults when `blend` is not set.
2. Why does it only apply to floating windows? It could be useful to make parts of the background see through, if the terminal or GUI supports it?
3. How do you even implement this in a GUI? The UI protocol still reports the highlight `blend` value when `winblend` is set to zero for a floating window and the raw `winblend` and `pumblend` values are not included in the protocol. I don't consider the `OptionSet` as a valid approach, since it will cause all kinds of synchronization issues. But of course, if the above two restrictions could be removed, then that's no longer a problem.

**Comment by bfredl** [comment]
we could clarify that this documented behavior is just for what the TUI is doing. Another GUI could of course do other things beyond what is possible in the TUI.

**Reply by fredizzimo** [reply]
It would be very nice to have a consistent meaning everywhere, so that plugin authors and color scheme creators don't need to implement different versions depending on if a GUI is used or not. Of course some GUIs might still want to do their own thing, but Neovide for example tries to match the TUI as close as possible. 

I don't see any reasons why the TUI couldn't implement my first suggestion of always using the blend value no matter what winblend or pumblend is set to. It's of course a breaking change, but I don't think many plugins relies on this behaviour currently.

I also think that the specification could be updated to allow it for non-floating windows, also in the terminal. I don't think any terminals currently support it. But the Kitty protocol for example already supports a fixed list of transparent colors, so I don't see why it couldn't be extended to support true color transparency as well. It might even be hackable with the existing image protocol.

A non breaking way of doing it, is to add `blendfg` and `blendbg`, to fully implement https://github.com/neovim/neovim/issues/24472 (I realize now that my initial question might have better fitted in that issue). And deprecate `blend`, but let it function like currently for the TUI, with undefined behaviour on GUIs.

BTW, I will also open an issue, regarding `default_colors_set`, which doesn't currently provide a blend value, so implementing this purely on the UI side is not fully possible, it works for all highlight groups except the `Normal` one.

**Reply by fredizzimo** [reply]
I updated #24472 with my suggestion
